### PR TITLE
Update user subscription cache, in place if possible

### PR DESCRIPTION
**Pull request by schwers**
:eyeglasses: @ajacksified

### CODE REVIEW

File: src/endpoints/v1.es6.js [hunk 1]
```diff
@@ -399,15 +399,59 @@ class APIv1Endpoint {
...
+
+        if (!subscribedRequestCache) { return };
+
+        // if there's not the default entry we'll assume something fancy
+        // is going on and not update the cache
+        const keys = subscribedRequestCache.keys();
+        if (keys.length !== 1) {
+          // if there's more keys just blow away the cache
+          subscribedRequestCache.reset();
+          return
```
Comment (ajacksified) [inline src/endpoints/v1.es6.js]: `;`

File: src/endpoints/v1.es6.js [hunk 2]
```diff
@@ -399,15 +399,59 @@ class APIv1Endpoint {
...
+        };
+
+        const key = keys[0];
+        const requestCache = subscribedRequestCache.get(key);
+        if (!requestCache) { return };
```
Comment (ajacksified) [inline src/endpoints/v1.es6.js]: `;`

File: src/endpoints/v1.es6.js [hunk 3]
```diff
@@ -399,15 +399,59 @@ class APIv1Endpoint {
...
+
+        let subreddits = requestCache.subreddits || [];
+        const updateId = options.id;
+        if (action === 'unsub') {
+          subreddits = subreddits.filter(function(id) {
+            return id !== updateId;
+          });
+        } else {
+          if (!subreddits.some(function(id) { return id === updateId })) {
```
Comment (ajacksified) [inline src/endpoints/v1.es6.js]: `;`

File: src/endpoints/v1.es6.js [hunk 4]
```diff
@@ -399,15 +399,59 @@ class APIv1Endpoint {
...
+            // Update the subscriber request cache to include or exclude
+            // the new subreddit if the request didn't fail
+            if (!Object.keys(body).length) {
+              this.subscriptions.updateSubscribedCache(options, json.action);
+            }
+
             return body;
           }, 'subreddits', json.sr, {
-            user_is_subscribed: true
+            user_is_subscriber: json.action === 'unsub' ? false : true,
```
Comment (ajacksified) [inline src/endpoints/v1.es6.js]: could this just be `user_is_subscriber: json.action === 'sub'`?

File: src/endpoints/v1.es6.js [hunk 5]
```diff
@@ -399,15 +399,59 @@ class APIv1Endpoint {
...
             reject('Subscription', options.model, valid);
           });
         }
+      },
+      updateSubscribedCache: function(options, action) {
+        if (!options.id || !options.id.length || !action.length) {
+          return;
+        }
+
+        const subscribedCacheKey = `${options.origin}/subreddits/mine/subscriber.json`
```
Comment (ajacksified) [inline src/endpoints/v1.es6.js]: `;`

**Comment by ajacksified** [comment]
A couple of nits, but otherwise looks good. :nail_care:

**Comment by schwers** [comment]
:haircut:

**Comment by schwers** [comment]
@ajacksified Something I was thinking of: what are your thoughts on clearing the frontpage cache when subscriptions happen?

**Comment by ajacksified** [comment]
@schwers makes sense, as the same query will return different information.

A way to abstract the logic, so the cache works automatically, would be to query by passing in an array of subreddit names (rather than *no* subreddit name), to get the same results as the front page; but I suspect the logic is different when you request the front page rather than `suba+subb+subc+subd...`.

**Comment by ajacksified** [comment]
:fish:, unless you want to add in clearing out the frontpage cache

**Comment by ajacksified** [comment]
re-:fish:ed

**Comment by ajacksified** [comment]
https://github.com/reddit/snoode/commit/7711530638c328a10f1da6aa11490ad588fd4426

**Pull request closed without merging**